[Clarenceeey] iP

Clarenceeey opened the pull request:
# DukePro
 > “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- Text-based
- Easy to learn
- ~~FAST~~_SUPER_ FAST to use!

All you need to do is:

1. Download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. Double-click it
3. Add your tasks
4. Let it manage your tasks for you :+1:

And its **FREE**!

Features:
- [X] Managing tasks 
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the first 5 lines of the `main` method:

```java
  public static void main (String[] args) {
      Ui ui = new Ui();
      Storage storage = new Storage(DIRPATH, FILEPATH);
      Parser parser = new Parser();
      TaskList taskList = new TaskList();

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Clarenceeey committed:
Rename chatbot

Clarenceeey committed:
Greet and exit

Clarenceeey committed:
Add echo

Clarenceeey committed:
Add task to list

Clarenceeey committed:
Add abstract class Task

Clarenceeey committed:
Add class ToDo

Clarenceeey committed:
Add marking tasks

Clarenceeey committed:
Added unmark functionality

Clarenceeey committed:
Add method to add ToDo tasks

Clarenceeey committed:
Add DeadLine class

Clarenceeey committed:
Add default toString for tasks

Clarenceeey committed:
Implement toString for DeadLine

Clarenceeey committed:
Implement toString for ToDo

Clarenceeey committed:
Implement adding tasks for ToDo and DeadLine

Clarenceeey committed:
Implement Event class

Clarenceeey committed:
Implement adding Event to tasklist

Clarenceeey committed:
Add test cases, .bat file to run tests

Clarenceeey committed:
New exception NoDateException created

Clarenceeey committed:
New exception to handle empty name

Clarenceeey committed:
Add throwing error into DeadLine

Clarenceeey committed:
ToDo now throws exception

Clarenceeey committed:
Removed default constructor in Task

Clarenceeey committed:
Event constructor throws exception

Clarenceeey committed:
Invalid dates and names handled in tasks

Clarenceeey committed:
Implement NoDateException

Clarenceeey committed:
String slicing handled in MrTracker

Clarenceeey committed:
Fix bug where error code also printed

Clarenceeey committed:
Add new constructor for InvalidNameException

Clarenceeey committed:
Fixed bug where wrong variable name used

Clarenceeey committed:
Renamed to InvalidDateException

Clarenceeey committed:
Rename exception to InvalidDateException

Clarenceeey committed:
Fixed bug where InvalidDateException is always thrown

Clarenceeey committed:
Provide messages for InvalidDateException

Clarenceeey committed:
InvalidDateException accepts messages

Clarenceeey committed:
Fixed bug where tasks not printed properly

Clarenceeey committed:
Base implementation of deleteTask

Clarenceeey committed:
Add deleting of tasks

Clarenceeey committed:
Use enum to check for cases

Clarenceeey committed:
Add base implementation for loadData

Clarenceeey committed:
Add abstract method in Task for tasks to inherit

Clarenceeey committed:
Implement toSave method and overloaded constructor

Clarenceeey committed:
Initial implementation of saving tasks to file

Clarenceeey committed:
Add checking for if directory and file exists

Clarenceeey committed:
Merge branch 'branch-Level-7'

Clarenceeey committed:
Saved time as LocalDate objects

Clarenceeey committed:
Date parsed as LocalDate

Clarenceeey committed:
Merge branch 'branch-Level-8'

Clarenceeey committed:
LocalDate object stored in Event class

Clarenceeey committed:
Fix formatting error in for LocalDate in Deadline

Clarenceeey committed:
Add Parser class

Clarenceeey committed:
Add PrefixString enum

Clarenceeey committed:
Put storage logic in Storage class

Clarenceeey committed:
Put tasklist logic in TaskList class

Clarenceeey committed:
Put user interaction logic in Ui class

Clarenceeey committed:
Move logic from MrTracker to other classes

Clarenceeey committed:
Merge branch 'A-MoreOOP'

Clarenceeey committed:
Put files in packages

Clarenceeey committed:
Merge branch 'A-Packages'

Clarenceeey committed:
Merge branch 'add-gradle-support'

Clarenceeey committed:
Changed package structure

Clarenceeey committed:
Change package structure again

Clarenceeey committed:
Add test folder, implemented ToDoTest

Clarenceeey committed:
Change package names

Clarenceeey committed:
Change string formatting issues

Clarenceeey committed:
Implement tests for DeadLine class

Clarenceeey committed:
Merge branch 'A-JUnit'

Clarenceeey committed:
Change build.gradle to set archiveBaseName

Clarenceeey committed:
Add javadocs to all methods and classes

Clarenceeey committed:
Change code to match coding standard

Clarenceeey committed:
Implement find method and fix formatting issues

Clarenceeey committed:
Merge branch 'branch-A-JavaDoc'

Clarenceeey committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/processes/MrTracker.java
#	src/main/java/processes/Parser.java
#	src/main/java/processes/Storage.java
#	src/main/java/processes/TaskList.java
#	src/main/java/processes/Ui.java
#	src/main/java/tasks/DeadLine.java
#	src/main/java/tasks/Event.java
#	src/main/java/tasks/Task.java
#	src/main/java/tasks/ToDo.java

Clarenceeey committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/processes/TaskList.java
#	src/main/java/processes/Ui.java

Clarenceeey committed:
Add tests for Parser and Storage classes

Clarenceeey committed:
Merge branch 'master' into A-JUnit

# Conflicts:
#	build.gradle
#	src/main/java/processes/Parser.java

Clarenceeey committed:
Add checkstyle configs and plugins

Clarenceeey committed:
Change style to match coding standard

Clarenceeey committed:
Merge branch 'branch-A-CheckStyle'

Clarenceeey committed:
Add files to be used for GUI

Clarenceeey committed:
Move resources folder to correct location

Clarenceeey committed:
Add exception TaskOutOfBounds if user index out of list bounds

Clarenceeey committed:
Change name of methods

Clarenceeey committed:
Put output handling code in Ui class

Clarenceeey committed:
Add extra input handling to welcome users

Clarenceeey committed:
Handle welcoming users

Clarenceeey committed:
Add delay before app terminates

Clarenceeey committed:
Merge branch 'branch-Level-10'

hj235 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * Unmarks the task by setting flag to false
+     */
+    public void unMark() {
```
Might it be better to name the function unmark (without a capital M) instead?
Since the 'unmark' is being used as a single word in this context.

hj235 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @return save format of task.
+     */
+    public abstract String toSave();
```
Could the name `toSave` be a little confusing as to the usage of this method?
I now see the relation to the `toString` method from the `Object` class, however at first glance it was a little confusing.
Perhaps an alternative name could give more clarity, such as `getSaveFormat` or `formatToSave`.

hj235 commented on src/main/java/processes/Ui.java:
```diff
@@ -0,0 +1,130 @@
+     * @param unmarkedTask The task that was just unmarked.
+     * @return The success message
+     */
+    public String showUnmarked(Task unmarkedTask) {
+        return "task\n  " + unmarkedTask.toString() + "\nis unMarked!";
+    }
```
Could it be better to uncapitalize the 'M' in 'unmarked' for the returned `String`?
The method name `showUnmarked` implies 'unmarked' is being treated as a single word, however in the returned `String`, the capitalization is instead 'unMarked'.

hj235 commented on src/main/java/processes/PrefixString.java:
```diff
@@ -0,0 +1,60 @@
+package processes;
+
+/**
+ * PrefixString enum that contains the inputs the programme can handle.
+ * Helps match user inputs to one of the predefined, valid inputs
+ */
```
It might be good to give a more detailed description of what exactly a `PrefixString` is. 
It might be confusing that some of the values of the enum objects have a space appended to the end, while some do not,

hj235 commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,126 @@
+            String taskName = inputString.substring(0, fromIndex).trim();
+            if (taskName.length() == 0) {
+                throw new InvalidTaskNameException();
+            }
+
+            if (inputString.contains("/to ")) {
+                int toIndex = inputString.indexOf("/to ");
+                if (toIndex < fromIndex) {
+                    throw new InvalidDateException("Wrong date order provided!");
+                }
+                String fromDate = inputString.substring(fromIndex + 6, toIndex);
+                String toDate = inputString.substring(toIndex + 4);
+                if (fromDate.length() == 0) {
+                    throw new InvalidDateException("From date not provided");
+                } else if (toDate.length() == 0) {
+                    throw new InvalidDateException("To date not provided");
+                }
+                this.name = taskName;
+                try {
+                    this.fromDate = LocalDate.parse(fromDate.trim());
+                } catch (DateTimeParseException ex) {
+                    throw new InvalidDateException("Invalid from date format given");
+                }
+
+                try {
+                    this.toDate = LocalDate.parse(toDate.trim());
+                } catch (DateTimeParseException ex) {
+                    throw new InvalidDateException("Invalid to date format given");
+                }
+
+                if (this.toDate.isBefore(this.fromDate)) {
+                    throw new InvalidDateException("To date is before from date");
+                }
+            } else {
+                throw new InvalidDateException("To date is not provided!");
+            }
+        } else {
+            throw new InvalidDateException("From date is not provided!");
+        }
+    }
```
I'm not sure I like how the line breaks were placed for this method. Perhaps placing a few more line breaks (e.g. in lines 46, 48, 53) could help with readability of the code by segmenting it into smaller sections.

hj235 reviewed (commented):
Looks good to me! 👍 
Some cosmetic changes requested.

Clarenceeey committed:
Add assertions

Assertions needed because certain assumptions are made in
some methods in the program that are not being checked for.

Let's add some assertions where these assumptions are made
to ensure the program does not break.

Clarenceeey committed:
Refactor code to follow code quality

Some parts of the code were difficult to read.

Let's refactor the code such that the happy path is prominent.
Add new method to get length of tasklist to prevent convoluted
methods to get the size.

Clarenceeey committed:
Use streams to filter tasklist

Using streams make filtering more streamlined and readable

Clarenceeey committed:
Merge pull request #1 from Clarenceeey/branch-A-Assertions

Add assertions

Clarenceeey committed:
Merge branch 'master' into branch-A-CodeQuality

Clarenceeey committed:
Merge branch 'master' into branch-A-Streams

Clarenceeey committed:
Merge pull request #2 from Clarenceeey/branch-A-CodeQuality

Refactor code to follow code quality

Clarenceeey committed:
Merge branch 'master' into branch-A-Streams

Clarenceeey committed:
Merge pull request #3 from Clarenceeey/branch-A-Streams

Use streams to filter tasklist

Clarenceeey committed:
CI configuration

Clarenceeey committed:
Tasks can now have tags

Clarenceeey committed:
Modify string representation of tasks to account for tags

Clarenceeey committed:
Fix formatting issues when printing

Clarenceeey committed:
Stop tracking data

Clarenceeey committed:
Add new command: tag

Clarenceeey committed:
Remove redundant method

Clarenceeey committed:
Update formatting of strings in tasks

Clarenceeey committed:
Initial implementation of addTags

Clarenceeey committed:
Update test code

Clarenceeey committed:
Implement method to add task and display success message

Clarenceeey committed:
Tags can now be removed from tasks

Clarenceeey committed:
Merge branch 'branch-C-tagging'

Clarenceeey committed:
Enforce SLAP in getResponse method

Clarenceeey committed:
Change padding and style

Clarenceeey committed:
Add exceptions to handle tagging

Clarenceeey committed:
Merge branch 'branch-A-MoreErrorHandling'

Clarenceeey committed:
Add new exception and update tests

Clarenceeey committed:
Fix minor bug and set title of the Ui

Clarenceeey committed:
Update README file and upload a picture of the Ui

Clarenceeey committed:
Update readme doc
